godot-macros/src/ffi_macros.rs

```diff
@@ -13,22 +13,28 @@ use crate::util::bail;
 use crate::ParseResult;
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote};
+use std::env;
+use std::sync::atomic::{AtomicU32, Ordering};
+
+static FUNCTION_COUNTER: AtomicU32 = AtomicU32::new(0);
 
 pub(super) fn wasm_declare_init_fn(input: TokenStream) -> ParseResult<TokenStream> {
     if !input.is_empty() {
         return bail!(input, "macro expects no arguments");
     }
 
-    // Create sufficiently unique identifier without entire `uuid` (let alone `rand`) crate dependency.
-    let a = unsafe { libc::rand() };
-    let b = unsafe { libc::rand() };
+    let crate_name = env::var("CARGO_PKG_NAME")
+        .expect("CARGO_PKG_NAME env var not found. This macro must be run by Cargo.")
+        .replace('-', "_"); // crate names may contain hyphens, but Rust identifiers must not.
```

Comment:
Could this not generate ambiguity? Can a crate with a hyphen and another with the same name but with underscores coexist?

One way to avoid that would be to replace __ with ___ first and then - with __

Reply:
not reasonably, because to use the crate in rust code, you have to do the same transformation.

E.g. if I import a crate ``do-stuff``, I can't use it as ``do-stuff::foo``, I have to use it like ``do_stuff::foo``

Reply:
Then we should probably be using `CARGO_CRATE_NAME` instead, I believe it already uses the canonical crate name with replacements[^1].

[^1]: https://doc.rust-lang.org/cargo/reference/environment-variables.html